File: RAG/create-embeddings.ts
#!/usr/bin/env node

import { ChromaClient } from 'chromadb';
import { GoogleGenAI } from '@google/genai';
import { readFileSync } from 'fs';
import { config } from 'dotenv';

config();

async function readDataFile(filePath: string): Promise<string[]> {
  try {
    const content = readFileSync(filePath, 'utf-8');
    const chunks = content.split('\n\n').map(chunk => chunk.trim()).filter(chunk => chunk);
    console.log(`Successfully read ${chunks.length} chunks from ${filePath}`);
    return chunks;
  } catch (error) {
    console.log(`Error reading file: ${error}`);
    return [];
  }
}

async function createVectorDatabase() {
  const client = new ChromaClient({ path: "http://localhost:8000" });
  const collectionName = "rag_documents";
  
  try {
    try {
      await client.deleteCollection({ name: collectionName });
      console.log("Deleted existing collection");
    } catch {}
    
    const collection = await client.createCollection({
      name: collectionName,
      metadata: { "hnsw:space": "cosine" }
    });
    console.log(`Created/got collection: ${collectionName}`);
    return { client, collection };
  } catch (error) {
    console.log(`Error creating vector database: ${error}`);
    return { client: null, collection: null };
  }
}

async function createEmbeddingsAndStore(chunks: string[], collection: any) {
  if (!chunks.length) {
    console.log("No chunks to process!");
    return;
  }

  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error("GEMINI_API_KEY not found in environment variables");
  }

  console.log("Creating embeddings with Gemini API...");
  const genAI = new GoogleGenAI({apiKey});

  const embeddings = await Promise.all(chunks.map(async chunk => {
    const result = await genAI.models.embedContent({model: "text-embedding-004", contents: [chunk]});
    return result.embeddings?.[0]?.values;
  }));

  const ids = chunks.map((_, i) => `chunk_${i}`);
  const metadatas = chunks.map((chunk, i) => ({ chunk_index: i, text_length: chunk.length }));

  console.log(`Storing ${chunks.length} embeddings in vector database...`);
  
  await collection.add({
    embeddings,
    documents: chunks,
    metadatas,
    ids
  });

  console.log("✅ Successfully stored all embeddings!");
}

async function main() {
  console.log("🚀 Starting RAG embedding creation process...");
  
  const dataFile = "data.txt";
  const chunks = await readDataFile(dataFile);

  if (!chunks.length) {
    console.log("❌ No data to process. Exiting.");
    return;
  }

  const { client, collection } = await createVectorDatabase();
  if (!collection) {
    console.log("❌ Failed to create vector database. Exiting.");
    return;
  }

  await createEmbeddingsAndStore(chunks, collection);

  console.log(`\n🎉 RAG setup complete!`);
  console.log(`📊 Processed ${chunks.length} text chunks`);
  console.log(`💾 Vector database stored in: ./chroma_db`);
  console.log(`🔍 Ready for queries! Run: npm run query 'your question here'`);
}

main().catch(console.error);

File: RAG-python/create_embeddings.py
#!/usr/bin/env python3
"""
Script to create embeddings from data.txt and store them in a local vector database (ChromaDB).
"""

import os
import chromadb
from sentence_transformers import SentenceTransformer
from typing import List
import re
# from google import genai

def read_data_file(file_path: str) -> List[str]:
    """Read and split the data file into chunks."""
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            content = file.read()
        
        # Split by double newlines to get paragraphs, then filter empty ones
        chunks = [chunk.strip() for chunk in content.split('\n\n') if chunk.strip()]
        
        print(f"Successfully read {len(chunks)} chunks from {file_path}")
        return chunks
    
    except FileNotFoundError:
        print(f"Error: File {file_path} not found!")
        return []
    except Exception as e:
        print(f"Error reading file: {e}")
        return []

def create_vector_database(): 
    """Initialize ChromaDB client and create collection."""
    # Create persistent client (stores data locally)
    client = chromadb.PersistentClient(path="./chroma_db")
    
    # Create or get collection
    collection_name = "rag_documents"
    try:
        # Try to delete existing collection to start fresh
        try:
            client.delete_collection(collection_name)
            print("Deleted existing collection")
        except:
            pass  # Collection doesn't exist, that's fine
        
        collection = client.create_collection(
            name=collection_name,
            metadata={"hnsw:space": "cosine"}  # Use cosine similarity
        )
        print(f"Created new collection: {collection_name}")
        return client, collection
    
    except Exception as e:
        print(f"Error creating vector database: {e}")
        return None, None

def create_embeddings_and_store(chunks: List[str], collection):
    """Create embeddings using SentenceTransformer and store in ChromaDB."""
    if not chunks:
        print("No chunks to process!")
        return
    
    print("Loading SentenceTransformer model...")
    # Use a good general-purpose embedding model
    model = SentenceTransformer('all-MiniLM-L6-v2')
    
    print("Creating embeddings...")
    embeddings = model.encode(chunks)
    print(embeddings)
    # Prepare data for ChromaDB
    ids = [f"chunk_{i}" for i in range(len(chunks))]
    metadatas = [{"chunk_index": i, "text_length": len(chunk)} for i, chunk in enumerate(chunks)]
    
    print(f"Storing {len(chunks)} embeddings in vector database...")
    
    # Add to collection
    collection.add(
        embeddings=embeddings.tolist(),
        documents=chunks,
        metadatas=metadatas,
        ids=ids
    )
    
    print("✅ Successfully stored all embeddings!")

def main():
    print("🚀 Starting RAG embedding creation process...")
    
    # Read the data file
    data_file = "data.txt"
    chunks = read_data_file(data_file)

    if not chunks:
        print("❌ No data to process. Exiting.")
        return
    
    # Create vector database
    client, collection = create_vector_database()
    if not collection:
        print("❌ Failed to create vector database. Exiting.")
        return
    
    # Create and store embeddings
    create_embeddings_and_store(chunks, collection)
    
    print(f"\n🎉 RAG setup complete!")
    print(f"📊 Processed {len(chunks)} text chunks")
    print(f"💾 Vector database stored in: ./chroma_db")
    print(f"🔍 Ready for queries! Run: python query_rag.py 'your question here'")

if __name__ == "__main__":
    main() 